### PR TITLE
Added NETCONF notification subscribe and processing capability

### PR DESCRIPTION
**Pull request by nhattrinh**
Added a NetconfNotification class which subclasses a Thread class to listen for notifications from the device as a result of establishing a subscription with a device. Added notify_wait into the Netconf class to trigger event, and added subscribe function in Netconf class to create the NetconfNotification thread, start listening for notifications and add the thread to active notifications.

At the start of the subscribe action, terminal will show the banner:
image

When listening for notifications, terminal will show the following:
image

When the subscribe action times out, the terminal will show:
image

Sample log: [job_output.log](https://github.com/CiscoTestAutomation/yang/files/5692559/job_output.log)

### CODE REVIEW

File: connector/src/yang/connector/netconf.py [hunk 1]
```diff
@@ -1,11 +1,17 @@
 """netconf.py module is a wrapper around the ncclient package."""
 
 import re
+import time
+from datetime import datetime as dt
```
Comment (LenzDong) [inline connector/src/yang/connector/netconf.py:5]: please remove duplicate import, line 12 already import `datetime`

File: connector/src/yang/connector/netconf.py [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+import atexit
+import logging
```
Comment (LenzDong) [inline connector/src/yang/connector/netconf.py:7]: please remove this duplicate import

**Comment by ahersi2-cisco** [comment]
@yuekyang can you please take a look at this PR as a final check? Thank you!

File: connector/src/yang/connector/netconf.py [hunk 3]
```diff
@@ -849,3 +890,64 @@ def _request(self, msg):
...
+
+    def run(self):
+        t1 = dt.now()
+        logger.info(banner('Starting subscribe stream'))
+        while not self.stopped():
+            t2 = dt.now()
+            td = t2 - t1
+            notif = self.device.take_notification(timeout=1)
+
+            logger.info('Listening for notifications from subscribe stream, {} seconds elapsed'.format(td.seconds))
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:934]: This line is better to stay with `t2 = dt.now()`.

File: connector/src/yang/connector/netconf.py [hunk 4]
```diff
@@ -849,3 +890,64 @@ def _request(self, msg):
                 logger.info('Timeout. No rpc-reply received.')
                 raise TimeoutExpiredError('ncclient timed out while waiting '
                                           'for an rpc-reply.')
+
+
+class NetconfNotification(Thread):
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:895]: How about calling it `NotificationThread` ?

File: connector/src/yang/connector/netconf.py [hunk 5]
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
             self, session=session, device_handler=device_handler,
             timeout=self.timeout)
 
+    active_notifications = {}
+
+    def subscribe(self, request, steps):
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:221]: `steps` is not used here. It can be removed.

File: connector/src/yang/connector/netconf.py [hunk 6]
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
             self, session=session, device_handler=device_handler,
             timeout=self.timeout)
 
+    active_notifications = {}
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:219]: Do you mean:
`self.active_notifications = {}`

File: connector/src/yang/connector/netconf.py [hunk 7]
```diff
@@ -210,6 +216,18 @@ def __init__(self, *args, **kwargs):
...
+
+    def subscribe(self, request, steps):
+        subscribe_thread = NetconfNotification(self, request=request)
+
+        if not self.connected:
+            self.connect()
+
+        subscribe_thread.start()
+
+        self.active_notifications[self] = subscribe_thread
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:229]: I'm not convinced to use `self` as a key of `self.active_notifications`. For one Netconf session, how many `subscribe_thread` are allowed?

File: connector/src/yang/connector/netconf.py [hunk 8]
```diff
@@ -414,6 +432,29 @@ def disconnect(self):
 
         self.session.close()
 
+    def notify_wait(self, steps):
```
Comment (yuekyang) [inline connector/src/yang/connector/netconf.py:435]: `steps` is a concept in pyATS test cases. Is it possible to avoid passing in `steps`?

**Review by yuekyang — COMMENTED** [review]
Thanks for the `class NetconfNotification`. I think it's useful.

File: connector/src/yang/connector/netconf.py (hunk 8 above)
Comment (nhattrinh) [reply connector/src/yang/connector/netconf.py:435]: No, I haven't seen another alternative from my current standpoint.

File: connector/src/yang/connector/netconf.py (hunk 7 above)
Comment (nhattrinh) [reply connector/src/yang/connector/netconf.py:229]: should I use a unique ID for the key of `self.active_notifications`? Currently for one NETCONF session, there is no limit on how many `subscribe_thread` are allowed.

Comment (nhattrinh) [reply connector/src/yang/connector/netconf.py]: I used a unique id to reference the notification thread

**Comment by nhattrinh** [comment]
I have addressed all the comments, please review my changes.

**Comment by nhattrinh** [comment]
Closed this PR and created a new PR (https://github.com/CiscoTestAutomation/yang/pull/28).

The new PR has addressed all the review comments in this PR.

**Pull request closed without merging**